# **GOAL:** To Identify **Customer Segments** For Online Retail With The Use Of **K-Means Clustering**.

---
## 1. IMPORTING LIBRARIES
---

In [298]:
# import base libraries 
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# import other libraries
from sklearn.preprocessing import StandardScaler
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from mpl_toolkits.mplot3d import Axes3D

---
## 2. LOADING THE DATASET
---

In [300]:
# laod the dataset
df = pd.read_excel('Online Retail.xlsx')
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

In [301]:
# look at the complete dataset
print(f"Shape of data: {df.shape}\n")

# info
df.info()

Shape of data: (541909, 8)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


**CONCLUSION:**
* There are 541909 rows and 8 columns before cleaning.
* 'InvoiceNo' consists of invoice ID of the transaction.
* 'StockCode' consists of stock ID.
* 'Description' consists of details about the stock being bought.
* 'Quantity' consists of the quantity bought.
* 'InvoiceDate' is the date and time of the invoice generation.
* 'UnitPrice' is the unit price of the product being bought.
* 'CustomerID' is the unique ID allotted to the customer.
* 'Country' specifies the country of the customer.

In [303]:
# description
df.describe(include="all").T

count   unique                                 top    freq  \
InvoiceNo    541909.0  25900.0                            573585.0  1114.0   
StockCode      541909     4070                              85123A    2313   
Description    540455     4223  WHITE HANGING HEART T-LIGHT HOLDER    2369   
Quantity     541909.0      NaN                                 NaN     NaN   
InvoiceDate    541909      NaN                                 NaN     NaN   
UnitPrice    541909.0      NaN                                 NaN     NaN   
CustomerID   406829.0      NaN                                 NaN     NaN   
Country        541909       38                      United Kingdom  495478   

                                      mean                  min  \
InvoiceNo                              NaN                  NaN   
StockCode                              NaN                  NaN   
Description                            NaN                  NaN   
Quantity                           9.55225             -80995.0   
InvoiceDate  2011-07-04 13:34:57.156386048  2010-12-01 08:26:00   
UnitPrice                         4.611114            -11062.06   
CustomerID                     15287.69057              12346.0   
Country                                NaN                  NaN   

                             25%                  50%                  75%  \
InvoiceNo                    NaN                  NaN                  NaN   
StockCode                    NaN                  NaN                  NaN   
Description                  NaN                  NaN                  NaN   
Quantity                     1.0                  3.0                 10.0   
InvoiceDate  2011-03-28 11:34:00  2011-07-19 17:17:00  2011-10-19 11:27:00   
UnitPrice                   1.25                 2.08                 4.13   
CustomerID               13953.0              15152.0              16791.0   
Country                      NaN                  NaN                  NaN   

                             max          std  
InvoiceNo                    NaN          NaN  
StockCode                    NaN          NaN  
Description                  NaN          NaN  
Quantity                 80995.0   218.081158  
InvoiceDate  2011-12-09 12:50:00          NaN  
UnitPrice                38970.0    96.759853  
CustomerID               18287.0  1713.600303  
Country                      NaN          NaN

---
## 3. DATA CLEANING
---

In [305]:
# check duplicate rows
df[df.duplicated()]

InvoiceNo StockCode                        Description  Quantity  \
517       536409     21866        UNION JACK FLAG LUGGAGE TAG         1   
527       536409     22866      HAND WARMER SCOTTY DOG DESIGN         1   
537       536409     22900    SET 2 TEA TOWELS I LOVE LONDON          1   
539       536409     22111       SCOTTIE DOG HOT WATER BOTTLE         1   
555       536412     22327  ROUND SNACK BOXES SET OF 4 SKULLS         1   
...          ...       ...                                ...       ...   
541675    581538     22068        BLACK PIRATE TREASURE CHEST         1   
541689    581538     23318     BOX OF 6 MINI VINTAGE CRACKERS         1   
541692    581538     22992             REVOLVER WOODEN RULER          1   
541699    581538     22694                       WICKER STAR          1   
541701    581538     23343       JUMBO BAG VINTAGE CHRISTMAS          1   

               InvoiceDate  UnitPrice  CustomerID         Country  
517    2010-12-01 11:45:00       1.25     17908.0  United Kingdom  
527    2010-12-01 11:45:00       2.10     17908.0  United Kingdom  
537    2010-12-01 11:45:00       2.95     17908.0  United Kingdom  
539    2010-12-01 11:45:00       4.95     17908.0  United Kingdom  
555    2010-12-01 11:49:00       2.95     17920.0  United Kingdom  
...                    ...        ...         ...             ...  
541675 2011-12-09 11:34:00       0.39     14446.0  United Kingdom  
541689 2011-12-09 11:34:00       2.49     14446.0  United Kingdom  
541692 2011-12-09 11:34:00       1.95     14446.0  United Kingdom  
541699 2011-12-09 11:34:00       2.10     14446.0  United Kingdom  
541701 2011-12-09 11:34:00       2.08     14446.0  United Kingdom  

[5268 rows x 8 columns]

In [306]:
# duplicate rows percentage
print(f"Percentage of duplicate rows:{(5268/len(df)) * 100}")

# NOTE: about 1%.

Percentage of duplicate rows:0.9721189350979592


In [307]:
# drop duplicates
df.drop_duplicates(inplace=True)

print("Info after dropping the duplicate rows:\n")
df.info()

Info after dropping the duplicate rows:

<class 'pandas.core.frame.DataFrame'>
Index: 536641 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    536641 non-null  object        
 1   StockCode    536641 non-null  object        
 2   Description  535187 non-null  object        
 3   Quantity     536641 non-null  int64         
 4   InvoiceDate  536641 non-null  datetime64[ns]
 5   UnitPrice    536641 non-null  float64       
 6   CustomerID   401604 non-null  float64       
 7   Country      536641 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 36.8+ MB


**CONCLUSION:**
* After dropping the duplicate rows, 536641 total rows remain.

In [309]:
# look at null values
print("The null values are:")
print(df.isna().sum(), "\n")

# look at null value percentage
print("The null value percentages are:")
df.isna().mean() * 100

The null values are:
InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135037
Country             0
dtype: int64 

The null value percentages are:


InvoiceNo       0.000000
StockCode       0.000000
Description     0.270945
Quantity        0.000000
InvoiceDate     0.000000
UnitPrice       0.000000
CustomerID     25.163377
Country         0.000000
dtype: float64

**CONCLUSION:**
* 'Description' and 'CustomerID' have null values.
* For 'Description' column, as null values are <1%, we will drop those rows.

In [311]:
# drop null value rows of 'Description' column
df = df.dropna(subset='Description')
print(df.info())

# look at the null values again
print("\nNull values after cleaning 'Description' column:")
print(df.isna().sum())

# look at the null value percentages again
print("\nNow the null value percentages are:")
df.isna().mean() * 100

<class 'pandas.core.frame.DataFrame'>
Index: 535187 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    535187 non-null  object        
 1   StockCode    535187 non-null  object        
 2   Description  535187 non-null  object        
 3   Quantity     535187 non-null  int64         
 4   InvoiceDate  535187 non-null  datetime64[ns]
 5   UnitPrice    535187 non-null  float64       
 6   CustomerID   401604 non-null  float64       
 7   Country      535187 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 36.7+ MB
None

Null values after cleaning 'Description' column:
InvoiceNo           0
StockCode           0
Description         0
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     133583
Country             0
dtype: int64

Now the null value percentages are:


InvoiceNo       0.000000
StockCode       0.000000
Description     0.000000
Quantity        0.000000
InvoiceDate     0.000000
UnitPrice       0.000000
CustomerID     24.960061
Country         0.000000
dtype: float64

**CONCLUSION:**
* Total rows now: 535187.
* Only 'CustomerID' column left to clean next.

In [313]:
# check all the unique elements in each column
for column in df.columns:
    print("*" * 30)
    print(f"The unique values in '{column}' is:")
    print(df[column].unique())
    print(f"The count of unique values in '{column}' is:")
    print(df[column].nunique())
    print("*" * 30, "\n")

******************************
The unique values in 'InvoiceNo' is:
[536365 536366 536367 ... 581585 581586 581587]
The count of unique values in 'InvoiceNo' is:
24446
****************************** 

******************************
The unique values in 'StockCode' is:
['85123A' 71053 '84406B' ... '90214U' '47591b' 23843]
The count of unique values in 'StockCode' is:
3958
****************************** 

******************************
The unique values in 'Description' is:
['WHITE HANGING HEART T-LIGHT HOLDER' 'WHITE METAL LANTERN'
 'CREAM CUPID HEARTS COAT HANGER' ... 'lost'
 'CREAM HANGING HEART T-LIGHT HOLDER' 'PAPER CRAFT , LITTLE BIRDIE']
The count of unique values in 'Description' is:
4223
****************************** 

******************************
The unique values in 'Quantity' is:
[     6      8      2     32      3      4     24     12     48     18
     20     36     80     64     10    120     96     23      5      1
     -1     50     40    100    192    432    144    2

**CONCLUSION:**
* There are thousands of unique values in each columns except 'Country', so we **can't** look at it manually.
* Also, 'Quantity' has negative values which needs to be further investigated.

In [315]:
# check unique products with negative quantities
negative_quantity_df = df[df['Quantity'] < 0]
negative_product_group = negative_quantity_df.groupby('Description')['Quantity'].sum()
print(negative_product_group, "\n")

# check the invoices of the associated quantity
negative_quantity_df = df[df['Quantity'] < 0]
negative_quantity_invoice_group = negative_quantity_df.groupby('InvoiceNo')['Quantity'].sum()
print(negative_quantity_invoice_group, "\n")

# seperating the invoices
# separate positive and negative quantities
normal_invoices = df[df['Quantity'] > 0]
return_invoices = df[df['Quantity'] < 0]

# display first few rows of normal and return invoices
print("*" * 100)
print("Normal Invoices:")
print(f"Total normal invoices: {len(normal_invoices)}")
print(normal_invoices[["InvoiceNo", "Description", "Quantity"]].head(), "\n")

print("*" * 100)
print("Return/Cancelled Invoices:")
print(f"Total return/cancel invoices: {len(return_invoices)}")
print(return_invoices[["InvoiceNo", "Description", "Quantity"]].head())

Description
20713                               -400
 50'S CHRISTMAS GIFT BAG LARGE        -2
 DOLLY GIRL BEAKER                    -7
 OVAL WALL MIRROR DIAMANTE            -5
 SET 2 TEA TOWELS I LOVE LONDON       -8
                                    ... 
wrongly coded-23343                 -800
wrongly marked carton 22804         -256
wrongly marked. 23343 in box       -3100
wrongly sold as sets                -600
wrongly sold sets                   -975
Name: Quantity, Length: 2110, dtype: int64 

InvoiceNo
537032       -30
537425       -20
537426       -35
537432       -43
538072       -13
           ...  
C581484   -80995
C581490      -23
C581499       -1
C581568       -5
C581569       -6
Name: Quantity, Length: 4310, dtype: int64 

****************************************************************************************************
Normal Invoices:
Total normal invoices: 525462
  InvoiceNo                          Description  Quantity
0    536365   WHITE HANGING HEART T-LIGHT H

In [316]:
# ratio of cancelled invoice vs normal invoice
(9725/525462)

0.01850752290365431

**CONCLUSION:**
* The normal invoices (525462) have integer values while the cancelled/returned invoice starts with 'C' (9725) . As it is very small in quantity (about ratio of 0.018), we will drop it.

In [318]:
# drop the negative quantity
df = df[df["Quantity"]>0]
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 525462 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    525462 non-null  object        
 1   StockCode    525462 non-null  object        
 2   Description  525462 non-null  object        
 3   Quantity     525462 non-null  int64         
 4   InvoiceDate  525462 non-null  datetime64[ns]
 5   UnitPrice    525462 non-null  float64       
 6   CustomerID   392732 non-null  float64       
 7   Country      525462 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 36.1+ MB


In [319]:
# look at the negative quantities again
df[df["Quantity"]<0]

# so, there isn't any negative quantities.

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

In [320]:
# drop the null values of 'CustomerID'
df = df.dropna(subset="CustomerID")

# look at the null values again
print(f"Checking for any null values: {df.isna().sum()}\n")

# shape of data
print(f"Now, shape of dataset: {df.shape}")

Checking for any null values: InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64

Now, shape of dataset: (392732, 8)


**CONCLUSION:**
* The 'CustomerID' column needs to be non-null for Customer Segmentation, so we will drop the null rows here which will reduce our data to about 75%(400000) which is still substantial. This will ensure that K means clustering analysis of the customers find valuable data in 'CustomerID'.
* Finally, data is cleaned and we can move to standardisation and other preprocessing.

---
## 4. DATA TYPE CONVERSION
---

In [323]:
# changing datatype of 'CustomerID' to relevant int64 type
df["CustomerID"] = df["CustomerID"].astype('int64')

# look at info again
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 392732 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    392732 non-null  object        
 1   StockCode    392732 non-null  object        
 2   Description  392732 non-null  object        
 3   Quantity     392732 non-null  int64         
 4   InvoiceDate  392732 non-null  datetime64[ns]
 5   UnitPrice    392732 non-null  float64       
 6   CustomerID   392732 non-null  int64         
 7   Country      392732 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(2), object(4)
memory usage: 27.0+ MB


In [324]:
# changing datatype of 'InvoiceNo'
df["InvoiceNo"] = df["InvoiceNo"].astype('int64')

# look at info
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 392732 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    392732 non-null  int64         
 1   StockCode    392732 non-null  object        
 2   Description  392732 non-null  object        
 3   Quantity     392732 non-null  int64         
 4   InvoiceDate  392732 non-null  datetime64[ns]
 5   UnitPrice    392732 non-null  float64       
 6   CustomerID   392732 non-null  int64         
 7   Country      392732 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(3), object(3)
memory usage: 27.0+ MB


In [325]:
# look at 0 unit price
print(len(df[df["UnitPrice"]==0]))
df[df["UnitPrice"]==0]


40


InvoiceNo StockCode                          Description  Quantity  \
9302       537197     22841         ROUND CAKE TIN VINTAGE GREEN         1   
33576      539263     22580         ADVENT CALENDAR GINGHAM SACK         4   
40089      539722     22423             REGENCY CAKESTAND 3 TIER        10   
47068      540372     22090              PAPER BUNTING RETROSPOT        24   
47070      540372     22553               PLASTERS IN TIN SKULLS        24   
56674      541109     22168        ORGANISER WOOD ANTIQUE WHITE          1   
86789      543599    84535B         FAIRY CAKES NOTEBOOK A6 SIZE        16   
130188     547417     22062  CERAMIC BOWL WITH LOVE HEART DESIGN        36   
139453     548318     22055   MINI CAKE STAND  HANGING STRAWBERY         5   
145208     548871     22162          HEART GARLAND RUSTIC PADDED         2   
157042     550188     22636   CHILDS BREAKFAST SET CIRCUS PARADE         1   
187613     553000     47566                        PARTY BUNTING         4   
198383     554037     22619            SET OF 6 SOLDIER SKITTLES        80   
279324     561284     22167           OVAL WALL MIRROR DIAMANTE          1   
282912     561669     22960             JAM MAKING SET WITH JARS        11   
285657     561916         M                               Manual         1   
298054     562973     23157           SET OF 6 NATIVITY MAGNETS        240   
314745     564651     23270     SET OF 2 CERAMIC PAINTED HEARTS         96   
314746     564651     23268  SET OF 2 CERAMIC CHRISTMAS REINDEER       192   
314747     564651     22955             36 FOIL STAR CAKE CASES        144   
314748     564651     21786                   POLKADOT RAIN HAT        144   
358655     568158      PADS           PADS TO MATCH ALL CUSHIONS         1   
361825     568384         M                               Manual         1   
379913     569716     22778                   GLASS CLOCHE SMALL         2   
395529     571035         M                               Manual         1   
420404     572893     21208          PASTEL COLOUR HONEYCOMB FAN         5   
436428     574138     23234        BISCUIT TIN VINTAGE CHRISTMAS       216   
436597     574175     22065       CHRISTMAS PUDDING TRINKET POT         12   
436961     574252         M                               Manual         1   
439361     574469     22385            JUMBO BAG SPACEBOY DESIGN        12   
446125     574879     22625                   RED KITCHEN SCALES         2   
446793     574920     22899         CHILDREN'S APRON DOLLY GIRL          1   
446794     574920     23480       MINI LIGHTS WOODLAND MUSHROOMS         1   
454463     575579     22437        SET OF 9 BLACK SKULL BALLOONS        20   
454464     575579     22089        PAPER BUNTING VINTAGE PAISLEY        24   
479079     577129     22464          HANGING METAL HEART LANTERN         4   
479546     577168         M                               Manual         1   
480649     577314     23407       SET OF 2 TRAYS HOME SWEET HOME         2   
485985     577696         M                               Manual         1   
502122     578841     84826       ASSTD DESIGN 3D PAPER STICKERS     12540   

               InvoiceDate  UnitPrice  CustomerID         Country  
9302   2010-12-05 14:02:00        0.0       12647         Germany  
33576  2010-12-16 14:36:00        0.0       16560  United Kingdom  
40089  2010-12-21 13:45:00        0.0       14911            EIRE  
47068  2011-01-06 16:41:00        0.0       13081  United Kingdom  
47070  2011-01-06 16:41:00        0.0       13081  United Kingdom  
56674  2011-01-13 15:10:00        0.0       15107  United Kingdom  
86789  2011-02-10 13:08:00        0.0       17560  United Kingdom  
130188 2011-03-23 10:25:00        0.0       13239  United Kingdom  
139453 2011-03-30 12:45:00        0.0       13113  United Kingdom  
145208 2011-04-04 14:42:00        0.0       14410  United Kingdom  
157042 2011-04-14 18:57:00        0.0       12457     Switzer

In [326]:
# drop the rows with 'UnitPrice'=0
df = df[df["UnitPrice"]>0]

# check again
df.sort_values(by="UnitPrice")

InvoiceNo     StockCode                 Description  Quantity  \
359871     568200          PADS  PADS TO MATCH ALL CUSHIONS         1   
279045     561226          PADS  PADS TO MATCH ALL CUSHIONS         1   
361741     568375  BANK CHARGES                Bank Charges         1   
157195     550193          PADS  PADS TO MATCH ALL CUSHIONS         1   
286813     562044         16045  POPART WOODEN PENCILS ASST       100   
...           ...           ...                         ...       ...   
374542     569382             M                      Manual         1   
406406     571751             M                      Manual         1   
422351     573077             M                      Manual         1   
422376     573080             M                      Manual         1   
173382     551697          POST                     POSTAGE         1   

               InvoiceDate  UnitPrice  CustomerID         Country  
359871 2011-09-25 14:58:00      0.001       16198  United Kingdom  
279045 2011-07-26 10:13:00      0.001       15618  United Kingdom  
361741 2011-09-26 17:01:00      0.001       13405  United Kingdom  
157195 2011-04-15 09:27:00      0.001       13952  United Kingdom  
286813 2011-08-02 10:24:00      0.040       16607  United Kingdom  
...                    ...        ...         ...             ...  
374542 2011-10-03 16:44:00   3155.950       15502  United Kingdom  
406406 2011-10-19 11:18:00   3949.320       12744       Singapore  
422351 2011-10-27 14:13:00   4161.060       12536          France  
422376 2011-10-27 14:20:00   4161.060       12536          France  
173382 2011-05-03 13:46:00   8142.750       16029  United Kingdom  

[392692 rows x 8 columns]

**CONCLUSION:**
* As the 0 unit price doesn't make much sense, and more than that, it is just 40 rows, so we dropped them.

In [328]:
# look at the dataset again
print(df.info(), "\n")

# look at the null values
print("Final check of null values:")
df.isna().sum()

<class 'pandas.core.frame.DataFrame'>
Index: 392692 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    392692 non-null  int64         
 1   StockCode    392692 non-null  object        
 2   Description  392692 non-null  object        
 3   Quantity     392692 non-null  int64         
 4   InvoiceDate  392692 non-null  datetime64[ns]
 5   UnitPrice    392692 non-null  float64       
 6   CustomerID   392692 non-null  int64         
 7   Country      392692 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(3), object(3)
memory usage: 27.0+ MB
None 

Final check of null values:


InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64

---
## 5. FEATURE ENGINEERING
---

**NOTE:** We can do RFM (Recency, Frequency and Monetary) analysis for which we need to create new features.

In [331]:
# create recency
latest_date = df["InvoiceDate"].max()
df["Recency"] = (latest_date - df["InvoiceDate"]).dt.days
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0     536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1     536365     71053                  WHITE METAL LANTERN         6   
2     536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3     536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4     536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  Recency  
0 2010-12-01 08:26:00       2.55       17850  United Kingdom      373  
1 2010-12-01 08:26:00       3.39       17850  United Kingdom      373  
2 2010-12-01 08:26:00       2.75       17850  United Kingdom      373  
3 2010-12-01 08:26:00       3.39       17850  United Kingdom      373  
4 2010-12-01 08:26:00       3.39       17850  United Kingdom      373

In [332]:
# create monetary (total money spent)
df["TotalPrice"] = df["Quantity"] * df["UnitPrice"]
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0     536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1     536365     71053                  WHITE METAL LANTERN         6   
2     536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3     536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4     536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  Recency  \
0 2010-12-01 08:26:00       2.55       17850  United Kingdom      373   
1 2010-12-01 08:26:00       3.39       17850  United Kingdom      373   
2 2010-12-01 08:26:00       2.75       17850  United Kingdom      373   
3 2010-12-01 08:26:00       3.39       17850  United Kingdom      373   
4 2010-12-01 08:26:00       3.39       17850  United Kingdom      373   

   TotalPrice  
0       15.30  
1       20.34  
2       22.00  
3       20.34  
4       20.34

In [333]:
# create frequency
frequency_df = df.groupby("CustomerID")["InvoiceNo"].nunique()
print("The count of invoice of each customer is:")
print(frequency_df)

# merging the frequency_df
df = pd.merge(df, frequency_df, on="CustomerID", how="left")

df.head()

The count of invoice of each customer is:
CustomerID
12346     1
12347     7
12348     4
12349     1
12350     1
         ..
18280     1
18281     1
18282     2
18283    16
18287     3
Name: InvoiceNo, Length: 4338, dtype: int64


InvoiceNo_x StockCode                          Description  Quantity  \
0       536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1       536365     71053                  WHITE METAL LANTERN         6   
2       536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3       536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4       536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  Recency  \
0 2010-12-01 08:26:00       2.55       17850  United Kingdom      373   
1 2010-12-01 08:26:00       3.39       17850  United Kingdom      373   
2 2010-12-01 08:26:00       2.75       17850  United Kingdom      373   
3 2010-12-01 08:26:00       3.39       17850  United Kingdom      373   
4 2010-12-01 08:26:00       3.39       17850  United Kingdom      373   

   TotalPrice  InvoiceNo_y  
0       15.30           34  
1       20.34           34  
2       22.00           34  
3       20.34           34  
4       20.34           34

In [334]:
# renaming the columns
df.rename({"InvoiceNo_x": "InvoiceNo", "InvoiceNo_y": "Frequency"}, inplace=True, axis=1)
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0     536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1     536365     71053                  WHITE METAL LANTERN         6   
2     536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3     536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4     536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  Recency  \
0 2010-12-01 08:26:00       2.55       17850  United Kingdom      373   
1 2010-12-01 08:26:00       3.39       17850  United Kingdom      373   
2 2010-12-01 08:26:00       2.75       17850  United Kingdom      373   
3 2010-12-01 08:26:00       3.39       17850  United Kingdom      373   
4 2010-12-01 08:26:00       3.39       17850  United Kingdom      373   

   TotalPrice  Frequency  
0       15.30         34  
1       20.34         34  
2       22.00         34  
3       20.34         34  
4       20.34         34

#### DROP THE IRRELEVANT COLUMNS
* From the dataset, 'InvoiceNo' will be dropped as it is transaction id and will not contribute towards behaviour of the customer.
* 'StockCode' will be dropped as it is product id and it has 3958 unique values which if onehotencoded will create problems in the algorithm.
* 'Description' will be dropped as it also has 4223 unique values and will suffer from the same problem.
* 'InvoiceDate' as it is of datetime datatype and already is used to extract 'Recency' column.
* 'Country' is also dropped for now as it is categorical and we are first concerned with RFM(Recency, Frequency, Monetary).

In [336]:
# look at the columns
df.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country', 'Recency', 'TotalPrice',
       'Frequency'],
      dtype='object')

In [337]:
# create new dataframe with the new features
feature_columns = ['Quantity', 'CustomerID', 'UnitPrice','Recency', 'TotalPrice', 'Frequency']

df_final = df[feature_columns].copy()
print(df_final.info())
df_final.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 392692 entries, 0 to 392691
Data columns (total 6 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   Quantity    392692 non-null  int64  
 1   CustomerID  392692 non-null  int64  
 2   UnitPrice   392692 non-null  float64
 3   Recency     392692 non-null  int64  
 4   TotalPrice  392692 non-null  float64
 5   Frequency   392692 non-null  int64  
dtypes: float64(2), int64(4)
memory usage: 18.0 MB
None


Quantity  CustomerID  UnitPrice  Recency  TotalPrice  Frequency
0         6       17850       2.55      373       15.30         34
1         6       17850       3.39      373       20.34         34
2         8       17850       2.75      373       22.00         34
3         6       17850       3.39      373       20.34         34
4         6       17850       3.39      373       20.34         34

## 6. FEATURE TRANSFORMATION

In [386]:
# feature scaling
rfm_feature = ['Quantity', 'UnitPrice', 'Recency', 'TotalPrice', 'Frequency']

# initialise scaler
scaler = StandardScaler()

# fit and transform
scaled = scaler.fit_transform(df_final[rfm_feature])

# convert back to dataframe
scaled_df = pd.DataFrame(scaled, columns=df_final[rfm_feature].columns)

scaled_df

Quantity  UnitPrice   Recency  TotalPrice  Frequency
0      -0.039446  -0.025893  1.968167   -0.023566   0.407949
1      -0.039446   0.011873  1.968167   -0.007366   0.407949
2      -0.028365  -0.016901  1.968167   -0.002030   0.407949
3      -0.039446   0.011873  1.968167   -0.007366   0.407949
4      -0.039446   0.011873  1.968167   -0.007366   0.407949
...          ...        ...       ...         ...        ...
392687 -0.006204  -0.102326 -1.341756   -0.039960  -0.396488
392688 -0.039446  -0.046125 -1.341756   -0.032245  -0.396488
392689 -0.050527   0.046043 -1.341756   -0.019388  -0.396488
392690 -0.050527   0.046043 -1.341756   -0.019388  -0.396488
392691 -0.056067   0.082012 -1.341756   -0.025013  -0.396488

[392692 rows x 5 columns]